src/Env.hs

```diff
@@ -170,6 +153,20 @@ emptyArgList = do
 
 ----------------------------------- Typechecker aux functions ------------------------------------------
 
+-- | Compute the transitive closure of list of pairs
+--   Useful for subtyping and equality aubtyping checkings
+transitiveClosure :: Eq a => [(a, a)] -> [(a, a)]
+transitiveClosure closure
+  | closure == closureAccum = closure
+  | otherwise                  = transitiveClosure closureAccum
+  where closureAccum =
+          nub $ closure ++ [(a, c) | (a, b) <- closure, (b', c) <- closure, b == b']
+
+-- | Return whether a closure is cyclic (a,b) and (b,a) appears in the closure
```

Comment:
If a closure is cyclic, then don't you just have to check for the presence of `(a, a)` in the closure?